torch/_refs/__init__.py

```diff
@@ -1658,49 +1658,135 @@ def item(a: TensorLikeType) -> NumberType:
+    dtype: Optional[torch.dtype] = None,
+    non_blocking: Optional[bool] = None,
+    copy: Optional[bool] = None,
+    memory_format: Optional[torch.memory_format] = None,
+    layout: Optional[torch.layout] = None,
+    pin_memory: Optional[bool] = None,
+):
+    arg_list = list(args)
+    arg0, *arg_list = (
+        arg_list
+        if arg_list
+        else [
+            None,
+        ]
+    )
+    arg0 = torch.device(arg0) if isinstance(arg0, str) else arg0
+    if isinstance(arg0, TensorLike):
+        # overload to `to.other`
+        device = arg0.device
+        dtype = arg0.dtype
+        layout = arg0.layout
+        # TODO: is_pinned is not currently supported in refs or fake_tensor
+        # https://github.com/pytorch/pytorch/issues/84925
+        # pin_memory = arg0.is_pinned()
+        # load positional arg_listuments:
+        if arg_list:
+            assert non_blocking is None
+            non_blocking, *arg_list = arg_list
+        if arg_list:
+            assert copy is None
+            copy, *arg_list = arg_list
+        if arg_list:
+            assert memory_format is None
+            memory_format, *arg_list = arg_list
+    elif isinstance(arg0, torch.dtype):
+        # overload to `to.dtype`
+        dtype = arg0
+        if arg_list:
+            assert non_blocking is None
+            non_blocking, *arg_list = arg_list
+        if arg_list:
+            assert copy is None
+            copy, *arg_list = arg_list
+        if arg_list:
+            assert memory_format is None
+            memory_format, *arg_list = arg_list
+    elif isinstance(arg0, torch.device):
+        # overload to `to.device`
+        device = arg0
+        if arg_list:
+            assert dtype is None
+            dtype, *arg_list = arg_list
+        if arg_list:
+            assert non_blocking is None
+            non_blocking, *arg_list = arg_list
+        if arg_list:
+            assert copy is None
+            copy, *arg_list = arg_list
+        if arg_list:
+            assert memory_format is None
+            memory_format, *arg_list = arg_list
+    else:
+        # overload to `to.dtype_layout` takes no positional arg_lists
+        assert arg0 is None
+    # assert that we have processed all positional argument
+    assert not arg_list
+    return device, dtype, non_blocking, copy, memory_format, layout, pin_memory
+
+
+def to(a: TensorLikeType, *args, **kwargs) -> TensorLikeType:
+    device, dtype, non_blocking, copy, memory_format, layout, pin_memory = _to_dispatch(
+        a, *args, **kwargs
+    )
+
+    if _to_will_alias(a, device, dtype, copy, memory_format, layout):
         return a
+    assert pin_memory is None
+    # short-circuit to `prims.convert_element_type` when `to` is just a dtype change
     if (
-        (copy is True or dtype != a.dtype)
-        and memory_format == torch.preserve_format
-        and non_blocking is False
+        (copy or dtype != a.dtype)
```

Comment:
This approach is fine for now but it'll be tricky to maintain. It'd be nicer if the changing properties were listed separately. Maybe `_to_dispatch` should return two dicts, and one of them is what's mutating? Maybe it should return only what's changing? Does the call to `empty_like` have to specify what isn't changing, to, or would it inherit that metadata, anyway?

Reply:
I think `empty_like` inherits whatever was from the original tensor. I added `canonicalize_to_arguments` that examines kwargs and attributes from input tensor and update `kwargs` to contain only what's changing.

Hopefully this makes it cleaner on the short-cut condition here.